Async fetch of shard started and store during allocation

kimchy opened the pull request:
Today, when a primary shard is not allocated we go to all the nodes to find where it is allocated (listing its started state). When we allocate a replica shard, we head to all the nodes and list its store to allocate the replica on a node that holds the closest matching index files to the primary.

Those two operations today execute synchronously within the GatewayAllocator, which means they execute in a sync manner within the cluster update thread. For large clusters, or environments with very slow disk, those operations will stall the cluster update thread, making it seem like its stuck.

Worse, if the FS is really slow, we timeout after 30s the operation (to not stall the cluster update thread completely). This means that we will have another run for the primary shard if we didn't find one, or we won't find the best node to place a shard since it might have timed out (listing stores need to list all files and read the checksum at the end of each file).

On top of that, this sync operation happen one shard at a time, so its effectively compounding the problem in a serial manner the more shards we have and the slower FS is...

This change moves to perform both listing the shard started states and the shard stores to an async manner. During the allocation by the GatewayAllocator, if data needs to be fetched from a node, it is done in an async fashion, with the response triggering a reroute to make sure the results will be taken into account. Also, if there are on going operations happening, the relevant shard data will not be taken into account until all the ongoing listing operations are done executing.

The execution of listing shard states and stores has been moved to their own respective thread pools (scaling, so will go down to 0 when not needed anymore, unbounded queue, since we don't want to timeout, just let it execute based on how fast the local FS is). This is needed sine we are going to blast nodes with a lot of requests and we need to make sure there is no thread explosion.

This change also improves the handling of shard failures coming from a specific node. Today, those nodes were ignored from allocation only for the single reroute round. Now, since fetching is async, we need to keep those failures around at least until a single successful fetch without the node is done, to make sure not to repeat allocating to the failed node all the time.

Note, if before the indication of slow allocation was high pending tasks since the allocator was waiting for responses, not the pending tasks will be much smaller. In order to still indicate that the cluster is in the middle of fetching shard data, 2 attributes were added to the cluster health API, indicating the number of ongoing fetches of both started shards and shard store.

Closes #9502

kimchy commented:
@martijnvg I need your help here, every once in a while, the test `IndicesStoreIntegrationTests#testShardActiveElseWhere` in network mode fails every once in a while. I don't full understand the test, since it creates fake started shards that causes the system to barf... . Can you have a look at it, I think the test should be rewritten.

martijnvg commented:
@kimchy yes, the test is confusing, because of inconsistencies... the idea behind this test is that the internal shard exist api is used to verify if we actually can the shards a node has on disk. 

By injecting fake started shards we let a node think that shards it has have been started on another node, It then verifies with the shard exist api if those shards do really exist on another node, in this test that isn't the case, so the node shouldn't delete its local shards.

kimchy commented:
@martijnvg yea, I gathered, it just creates an unexpected state in ES side. I am working on it now to see how far I go, will keep you posted.

kimchy commented:
@martijnvg I just rebased against master and pushed a fix for this (I think), where failed shards should be ignored for the next successful fetch phase, lets see how it goes

kimchy commented:
I found the problem in the `IndicesStoreIntegrationTests#testShardActiveElseWhere` test, pushed a fix here: https://github.com/elastic/elasticsearch/commit/ba20d4b6ba3902b97e0691e7fd965c0c0f684f42

imotov commented on src/test/java/org/elasticsearch/test/ElasticsearchIntegrationTest.java:
```diff
@@ -855,8 +855,14 @@ public void waitNoPendingTasksOnAll() throws Exception {
+                    assertThat("client " + client  + " still has fetch shard store", clusterHealth.getNumberOfFetchShardStore(), equalTo(0));
                     PendingClusterTasksResponse pendingTasks = client.admin().cluster().preparePendingClusterTasks().setLocal(true).get();
                     assertThat("client " + client + " still has pending tasks " + pendingTasks.prettyPrint(), pendingTasks, Matchers.emptyIterable());
+                    clusterHealth = client.admin().cluster().prepareHealth().setLocal(true).get();
```
I might be missing something but why cannot we just use the number of pending tasks from the first health response and get rid of the pending tasks and the second health requests? If there is a reason for it, I think it might be worth adding it as a comment.

kimchy replied:
I kept it as is, to keep the assert message above listing all the pending tasks, compared to just giving a number

imotov commented on src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -220,6 +234,11 @@ public void readFrom(StreamInput in) throws IOException {
                 validationFailures.add(in.readString());
             }
         }
+
```
The tag on PR says that it goes only to 2.0.0. Is this PR going to be back-ported to v1.6.0 as well?

kimchy replied:
I am going to try and see if this can be back ported to 1.6, if not, I will remove it afterwards

imotov commented on src/test/java/org/elasticsearch/indices/store/IndicesStoreIntegrationTests.java:
```diff
@@ -19,6 +19,7 @@
 
 package org.elasticsearch.indices.store;
 
+import com.carrotsearch.randomizedtesting.annotations.Repeat;
```
Testing leftover.

kimchy replied:
right, will remove

imotov commented on src/test/java/org/elasticsearch/cluster/MinimumMasterNodesTests.java:
```diff
@@ -126,29 +122,26 @@ public boolean apply(Object obj) {
         logger.info("--> starting the previous master node again...");
         internalCluster().startNode(settings);
 
+        ensureGreen();
```
Not sure I fully understand why are these ensureGreen were added. It seems that this ensureGreen is not needed since we are waiting for green in the next line and some other ensureGreen are either not needed (since we wait for number of active shards right after it) or they are making the tests less "evil" so to speak.

kimchy replied:
I added it to get a better message if we fail to wait for green, we anyhow wait for green in the next line like you observed, but we miss on better logging if we don't reach green. I want to improve this options and allow to specify number of nodes to the test framework, but prefer to do it in a different change

imotov commented:
@kimchy I left a few minor comments, otherwise it looks good to me. However, I don't have an intimate knowledge of some aspects of the allocation code that might be affected here (i.e. shadow replicas). So, I would feel much better if somebody else who worked with this code recently took another look. @dakrone, maybe?

kimchy commented:
@imotov removed the unneeded import, and responded to your comments. @dakrone it would be great if you can take a look

s1monw commented on src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -67,12 +70,15 @@
     /** needed for plugins BWC */
     public ClusterHealthResponse(String clusterName, String[] concreteIndices, ClusterState clusterState) {
-        this(clusterName, concreteIndices, clusterState, -1);
+        this(clusterName, concreteIndices, clusterState, -1, -1, -1);
```
can we have a constant instead of `-1` with a name?

kimchy replied:
since its only used for backward comp., I don't think we need to?

s1monw commented on src/main/java/org/elasticsearch/action/admin/cluster/health/ClusterHealthResponse.java:
```diff
@@ -57,6 +58,8 @@
     int initializingShards = 0;
     int unassignedShards = 0;
     int numberOfPendingTasks = 0;
+    int numberOfFetchShardStarted = 0;
```
I really can't tell what these are is this the total or the inflight? what did this fetch teh entire shard or only some info?

s1monw replied:
I also wonder if we can tell how many we still need to do if this makes sense?

kimchy replied:
yea, its the in flight ones, I will rename it to in flight and keep it simple and not separate to 2 types

s1monw commented on src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -0,0 +1,522 @@
+    /**
+     * Returns the number of async fetches that are currently ongoing.
+     */
+    public synchronized int getNumberOfAsyncFetches() {
```
can we call this `getNumberInFlightFetches` or so ?

kimchy replied:
yes! :)

s1monw commented on src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -0,0 +1,522 @@
+        }
+
+        // if we are still fetching, return null to indicate it
+        if (hasAnyNodeFetching(shardCache) == true) {
```
i think ` == true` can be skipped

kimchy replied:
I thought we do `== true` for readability?

s1monw commented on src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -0,0 +1,522 @@
+                // it next time we fetch the data
+                NodeEntry<T> nodeEntry = new NodeEntry<>(failedShard.currentNodeId());
+                nodeEntry.markAsFetching();
+                nodeEntry.doneFetching(new ElasticsearchException("marker for shard failed"));
```
what is a marker?

kimchy replied:
will have a better message

s1monw commented on src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -0,0 +1,522 @@
+     */
+    public synchronized int getNumberOfAsyncFetches() {
+        int count = 0;
+        for (Map<String, NodeEntry<T>> shard : cache.values()) {
```
can't we have this calculated on the fly?

kimchy replied:
it is calculated on the fly when calling this method. Do you mean keep it counted as things change? I prefer to keep it simple here since its not heavy calculation and its only really relevant during the heavy allocation phase of the cluster

s1monw commented on src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -0,0 +1,522 @@
+     * Schedules a reroute on the cluster update thread.
+     */
+    protected void reroute(ShardId shardId, String reason) {
+        clusterService.submitStateUpdateTask("done_fetch_shard_data for shard " + shardId + ", reasons (" + reason + ")", Priority.HIGH, new ClusterStateUpdateTask() {
```
can we factor this out of this class and remove the useless `@Inject` from the `AsyncShardFetch.Started` &  `AsyncShardFetch.Store`, make the `GatewayAllocator` instantiate these classes directly. We can then have a simple `OnShardFetched` callback that we can pass as a ctor argument that does the reroute for us. Similarly, the `GatewayAllocator` should also register the cluster change listener and if you ask me I'd drop all this generation thing  and replace the entire cache if we are getting a new master in the `GatewayAllocator`.

kimchy replied:
sure, will do and see how it goes

s1monw commented on src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -0,0 +1,522 @@
+            fetching = true;
+        }
+
+        public void doneFetching(T value) {
```
since tis is really an internal class do we really provide public access to all these mutators?

kimchy replied:
sure, will make those package private

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -266,6 +272,9 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
         unassignedIterator = routingNodes.unassigned().iterator();
         while (unassignedIterator.hasNext()) {
             MutableShardRouting shard = unassignedIterator.next();
+            if (shard.primary()) {
```
;)

s1monw commented on src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -284,18 +293,28 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
             if (!canBeAllocatedToAtLeastOneNode) {
+                // still fetching data, remove from the unassigned, and try the next
+                unassignedIterator.remove();
+                routingNodes.ignoredUnassigned().add(shard);
```
can we assign ` routingNodes.ignoredUnassigned()` to a variable please? it's used in some places and reads just easier

kimchy replied:
I will do this in another change, I want to change the allocator as little as possible with this change to easily backport

s1monw commented on src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -258,6 +258,11 @@ public Node start() {
         injector.getInstance(RestController.class).start();
+
+        // TODO hack around circular dependecncies problems
+        injector.getInstance(AsyncShardFetch.Started.class).setReallocation(injector.getInstance(ClusterService.class), injector.getInstance(AllocationService.class));
```
this is super ugly I think `AsyncShardFetch.Started` and `AsyncShardFetch.Store` should be an impl detail of `GatewayAllocator` no need to bind this or anything

kimchy replied:
need to do this to get away from circular dep problems, but will move it to `GatewayAllocator` so it won't be those async fetches and encapsulated in the allocator itself

s1monw commented:
I like this change - it really touches the problem we have today with this code and fixes it. Yet, I think we need to work on some code internals and naming before we get this in. left some comments.

dakrone commented on src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -0,0 +1,522 @@
+            this.fetching = false;
+        }
+
+        public void doneFetching(Throwable failure) {
```
What do you think of the name `finish(...)`? `doneFetching` is a little boolean sounding to me

dakrone commented on src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -0,0 +1,522 @@
+    /**
+     * An implementation of async shard fetch that list shard stores.
+     */
+    public static class Store extends AsyncShardFetch<TransportNodesListShardStoreMetaData.NodeStoreFilesMetaData> {
```
I think `Store` and `Started` should not be internal classes, it makes the `AsyncShardFetch` file smaller and easier to digest if they are moved to their own files.

kimchy replied:
I disagree :)

dakrone commented on src/main/java/org/elasticsearch/gateway/GatewayAllocator.java:
```diff
@@ -62,43 +51,29 @@
-            cachedStores.remove(shardRouting.shardId());
-            cachedShardsState.remove(shardRouting.shardId());
-        }
+        asyncFetchStarted.applyStartedShards(allocation);
```
Maybe add a comment here about what this is doing without having to jump to the source? (removing started shards from the caches)

kimchy replied:
repeat the doc from the javadoc? I feel like its excessive doc'ing and now there will be 3 places to sync, the doc here, the javadoc and the code?

dakrone commented on src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -125,6 +127,8 @@ public ThreadPool(Settings settings) {
                 .put(Names.SNAPSHOT, settingsBuilder().put("type", "scaling").put("keep_alive", "5m").put("size", halfProcMaxAt5).build())
                 .put(Names.OPTIMIZE, settingsBuilder().put("type", "fixed").put("size", 1).build())
+                .put(Names.FETCH_SHARD_STARTED, settingsBuilder().put("type", "scaling").put("keep_alive", "5m").put("size", availableProcessors * 2).build())
+                .put(Names.FETCH_SHARD_STORE, settingsBuilder().put("type", "scaling").put("keep_alive", "5m").put("size", availableProcessors * 2).build())
```
It looks like the shard fetching code correctly handles rejections on these threadpools, so do you think about setting a good default `queue_size` for these, to avoid any potential memory issues?

dakrone replied:
Ahh, I see the "unbounded queue, since we don't want to timeout, just let it execute based on how fast the local FS is)" now in the description, since you reschedule on rejection though, isn't that handled?

kimchy replied:
I put the handling rejection in place just in case we end up moving to fixed TP with bounded queue

kimchy replied:
but I don't think we should

kimchy commented:
@s1monw I pushed another round

s1monw commented on src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -0,0 +1,488 @@
+ * and once the results are back, it makes sure to schedule a reroute to make sure those results will
+ * be taken into account.
+ */
+public abstract class AsyncShardFetch<T extends NodeOperationResponse> extends AbstractComponent {
```
can we implement `Closeable` and use an AtomicBoolean to signal it's closed I like the `if (closed.compareAndSet(false, true))` pattern

s1monw commented on src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -0,0 +1,488 @@
+
+        @Override
+        protected void asyncFetch(final ShardId shardId, final String[] nodesIds, MetaData metaData) {
+            action.list(shardId, false, nodesIds, new ActionListener<TransportNodesListShardStoreMetaData.NodesStoreFilesMetaData>() {
```
can we just pass the action to the `AsyncShardFetch` class instead of subclassing this seems so close, maybe the two actions can share a common interface

s1monw commented on src/main/java/org/elasticsearch/gateway/AsyncShardFetch.java:
```diff
@@ -0,0 +1,488 @@
+    /**
+     * Schedules a reroute on the cluster update thread.
+     */
+    protected void reroute(ShardId shardId, String reason) {
```
this method seems to be wrong here the class is concerned with fetching data in an async fashion, I think what we do with it or what we call as a result should just be somewhere else. I wonder if we can just have an ActionListener that we pass to execute this onSuccess and onFailure since those seem to be the two mode we have.

kimchy commented:
@s1monw I pushed another round, tell me what you think...

s1monw commented:
LGTM great stuff @kimchy

kimchy commented:
thanks @s1monw, there were a few failures on CI for this pull which I wasn't happy with, I believe I tracked it down to not handling properly (as in, the same way as previously) the case of failed shards. I changed the code to have AsyncShardFetch to be shard level, and cleaned it up from having to mark the node as failed. I will let CI run a bit on it and see how it goes
